drivers/pwm/pwm_nrfx.c

```diff
@@ -0,0 +1,202 @@
+		nrf_pwm_configure(pconfig->pwm.p_registers,
+				  pconfig->config.base_clock, pconfig->config.count_mode,
+				  pdata->top_value);
+	}
+
+	if (pulse_cycles > pdata->top_value) {
+		return -EINVAL;
+	}
+
+	pdata->current[pwm] = pulse_cycles;
+
+	return 0;
+}
+
+static int pwm_nrfx_get_cycles_per_sec(struct device *dev, u32_t pwm,
+				       u64_t *cycles)
+{
+	const struct pwm_config *pconfig = (const struct pwm_config *)dev->config->config_info;
+
+	switch (pconfig->config.base_clock) {
+	case NRF_PWM_CLK_16MHz:
+		*cycles = 16ul * 1000ul * 1000ul;
+		break;
+	case NRF_PWM_CLK_8MHz:
+		*cycles = 8ul * 1000ul * 1000ul;
+		break;
+	case NRF_PWM_CLK_4MHz:
+		*cycles = 4ul * 1000ul * 1000ul;
+		break;
+	case NRF_PWM_CLK_2MHz:
+		*cycles = 2ul * 1000ul * 1000ul;
+		break;
+	case NRF_PWM_CLK_1MHz:
+		*cycles = 1ul * 1000ul * 1000ul;
+		break;
+	case NRF_PWM_CLK_500kHz:
+		*cycles = 500ul * 1000ul;
+		break;
+	case NRF_PWM_CLK_250kHz:
+		*cycles = 250ul * 1000ul;
+		break;
+	case NRF_PWM_CLK_125kHz:
+		*cycles = 125ul * 1000ul;
+		break;
+	}
+	return 0;
+}
+
+static const struct pwm_driver_api pwm_nrfx_drv_api_funcs = {
+	.pin_set = pwm_nrfx_pin_set,
+	.get_cycles_per_sec = pwm_nrfx_get_cycles_per_sec,
+};
+
+static int pwm_nrfx_init(struct device *dev)
+{
+	const struct pwm_config *pconfig = (const struct pwm_config *)dev->config->config_info;
+	struct pwm_data *pdata = dev->driver_data;
+
+	if (nrfx_pwm_init(&pconfig->pwm, &pconfig->config, NULL) == NRFX_SUCCESS) {
+		pdata->top_value = pconfig->config.top_value;
+		for (int ch = 0; ch < NRF_PWM_CHANNEL_COUNT; ch++) {
+			pdata->current[ch] = 0;
+		}
+		nrfx_pwm_simple_playback(&pconfig->pwm, &pdata->seq, 1, NRFX_PWM_FLAG_LOOP);
+		return 0;
+	} else {
+		return -1;
+	}
+}
+
+#define NRFX_PWM_DEVICE(idx)													      \
+	static const struct pwm_config pwm_nrfx_##idx##_config = {								      \
+		.pwm = NRFX_PWM_INSTANCE(idx),											      \
+		.config = {													      \
+			.output_pins = {											      \
+				CONFIG_PWM_##idx##_NRFX_CH0_PIN | (CONFIG_PWM_##idx##_NRFX_CH0_INVERTED ? NRFX_PWM_PIN_INVERTED : 0), \
+				CONFIG_PWM_##idx##_NRFX_CH1_PIN | (CONFIG_PWM_##idx##_NRFX_CH1_INVERTED ? NRFX_PWM_PIN_INVERTED : 0), \
+				CONFIG_PWM_##idx##_NRFX_CH2_PIN | (CONFIG_PWM_##idx##_NRFX_CH2_INVERTED ? NRFX_PWM_PIN_INVERTED : 0), \
+				CONFIG_PWM_##idx##_NRFX_CH3_PIN | (CONFIG_PWM_##idx##_NRFX_CH3_INVERTED ? NRFX_PWM_PIN_INVERTED : 0), \
+			},													      \
+			.irq_priority = NRFX_PWM_DEFAULT_CONFIG_IRQ_PRIORITY,							      \
+			.base_clock = (nrf_pwm_clk_t)CONFIG_PWM_##idx##_NRFX_CLOCK_PRESCALER,					      \
+			.count_mode = NRF_PWM_MODE_UP,										      \
+			.top_value = NRFX_PWM_DEFAULT_CONFIG_TOP_VALUE,								      \
+			.load_mode = NRF_PWM_LOAD_INDIVIDUAL,									      \
+			.step_mode = NRF_PWM_STEP_TRIGGERED,									      \
+		}														      \
+	};															      \
+	static struct pwm_data pwm_nrfx_##idx##_data = {									      \
```

Comment:
Adding `NRFX_PWM_PIN_INVERTED` to a pin number is not enough to make a pin inverted, since it only defines GPIO configuration of the pin. For an inverted channel you need to set bit 15 in the value supplied for this channel (see [doc](https://github.com/zephyrproject-rtos/zephyr/blob/master/ext/hal/nordic/nrfx/hal/nrf_pwm.h#L244)).
Therefore, I propose to initialize the channels here with 0 for normal and 0x8000 for inverted ones, and keep bit 15 untouched while setting a new value for a channel in `pwm_nrfx_pin_set`.

Reply:
I'll make the change, but from my testing, it seems like if bit 15 is set its normal, not set is inverted.  It's not really clear in that doc.

Reply:
You're right. I should be as you wrote - bit 15 set for normal (i.e. pin at 0 when idle, switching to 1 at compare match) channels and not set for inverted ones.
My bad, sorry.